### PR TITLE
Optimize boxing for var.GetType()

### PR DESCRIPTION
**Pull request by EgorBo**
```csharp
static Type Foo(int x) => x.GetType(); // to => typeof(int)
```
Currently [boxes](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBLANgHwAEAmARgFgAoKw0pAAhPoGEqBvK+rxu+gFQCeABxj0A4jAyCRACmwA7DPQQBKegF4AfMoB0EqcJgyVAbioBfIA===) `int x` in order to invoke `object.GetType`:
```asm
   0:   53                      push   %rbx
   1:   89 fb                   mov    %edi,%ebx
   3:   48 b8 70 01 c8 01 00    movabs $0x1c80170,%rax
   a:   00 00 00
   d:   bf 70 df d4 01          mov    $0x1d4df70,%edi
  12:   be 14 00 00 00          mov    $0x14,%esi
  17:   ff 10                   callq  *(%rax)
  19:   89 58 10                mov    %ebx,0x10(%rax)
  1c:   48 8b 00                mov    (%rax),%rax
  1f:   48 8b 40 18             mov    0x18(%rax),%rax
  23:   5b                      pop    %rbx
  24:   c3                      retq
```
This PR optimizes it to:
```asm
   0:   48 b8 30 cb 93 15 49    movabs $0x7f491593cb30,%rax
   7:   7f 00 00
   a:   c3                      retq
```
Issue: https://github.com/mono/mono/issues/17858

Jit-diff:
```
Total diff for System.Private.CoreLib.dll.so.dasm: -8621 bytes
```

### CODE REVIEW

**Comment by EgorBo** [comment]
@monojenkins build failed

File: mono/mini/method-to-ir.c [hunk 1]
```diff
@@ -8940,6 +8940,30 @@ mono_method_to_ir (MonoCompile *cfg, MonoMethod *method, MonoBasicBlock *start_b
...
+			//   call object::GetType()
+			//
+			guint32 gettype_token;
+			if ((ip = il_read_call(next_ip, end, &gettype_token)) && ip_in_bb (cfg, cfg->cbb, ip)) {
+				MonoMethod* gettype_method = mini_get_method (cfg, method, gettype_token, NULL, generic_context);
+				if (!strcmp (gettype_method->name, "GetType") && gettype_method->klass == mono_defaults.object_class) {
+					mono_class_init_internal(klass);
+					MonoType *klass_type = m_class_get_byval_arg (klass);
+					if (!mono_class_is_open_constructed_type (klass_type) && !cfg->compile_aot) { // FIXME: implement for AOT
+						// Emit `typeof(T)`
```
Comment (vargaz) [inline mono/mini/method-to-ir.c]: There is a
								EMIT_NEW_TYPE_FROM_HANDLE_CONST (cfg, ins, m_class_get_image (tclass), m_class_get_type_token (tclass), generic_context);
runtime constant which will work for aot as well.
It should be used like this:
if (mono_class_get_checked (m_class_get_image (tclass), m_class_get_type_token (tclass), error) == tclass)
								EMIT_NEW_TYPE_FROM_HANDLE_CONST (cfg, ins, m_class_get_image (tclass), m_class_get_type_token (tclass), generic_context);
To filter out classes which have no type token like arrays.

**Comment by EgorBo** [comment]
@monojenkins build failed

**Comment by EgorBo** [comment]
@monojenkins build Linux WebAssembly

File: mcs/class/corlib/Test/System.Reflection/FieldInfoTest.cs [hunk 2]
```diff
@@ -1356,6 +1356,7 @@ struct TestFields {
 		}
 
 		[Test]
+		[Category ("NotWorking")] // https://github.com/mono/mono/issues/18364
```
Comment (akoeplinger) [inline mcs/class/corlib/Test/System.Reflection/FieldInfoTest.cs]: looks like #18364 was fixed, can you try reenabling the test?

Comment (EgorBo) [reply mcs/class/corlib/Test/System.Reflection/FieldInfoTest.cs]: oh, indeed. Fixed.

**Comment by EgorBo** [comment]
@monojenkins build Linux AArch64 FullAOT+Interpreter+LLVM

**Comment by lewing** [comment]
@monojenkins build failed

**Comment by lewing** [comment]
@monojenkins build failed

**Comment by EgorBo** [comment]
@monojenkins build failed

**Pull request merged**